internal/charmstore/migrations.go

```diff
@@ -222,3 +227,93 @@ func populateWriteACL(db StoreDatabase) error {
 	logger.Infof("%d base entities updated", counter)
 	return nil
 }
+
+func populatePromulgatedEntities(db StoreDatabase) error {
+	baseEntities := db.BaseEntities()
+	entities := db.Entities()
+
+	// 1. Update all base entities to have a promulgated value of the correct form.
+	_, err := baseEntities.UpdateAll(
+		bson.D{{"promulgated", bson.D{{"$ne", 1}}}},
+		bson.D{{"$set", bson.D{{"promulgated", -1}}}},
+	)
+	if err != nil {
+		return errgo.Notef(err, "cannot set promulgated to initial false value")
+	}
+
+	// 2. Update entities with users with their promulgated URL.
+	//
+	// Note that we order the query results by revision, so when
+	// we've traversed all the entities, owner[name] will hold the
+	// owner of the latest revision of each promulgated charm with
+	// the given name.
+	owners := make(map[string]string)
+	iter := entities.Find(bson.D{{"user", ""}}).Select(bson.D{
+		{"_id", 1},
+		{"baseurl", 1},
+		{"blobhash", 1},
+		{"extrainfo.bzr-owner", 1},
+	}).Sort("name", "revision").Iter()
+	var entity mongodoc.Entity
+	for iter.Next(&entity) {
+		var user string
+		if err := json.Unmarshal([]byte(entity.ExtraInfo["bzr-owner"]), &user); err != nil {
+			return errgo.Notef(err, "cannot unmarshal user from extra-info")
+		}
+		if user == "" {
+			return errgo.Newf("no user for %q", entity.URL)
+		}
+		owners[entity.URL.Name] = user
+		err := entities.Update(
+			bson.D{
+				{"user", user},
+				{"name", entity.URL.Name},
+				{"series", entity.URL.Series},
+				{"blobhash", entity.BlobHash},
+				{"promulgated-url", bson.D{{"$exists", false}}},
+			},
+			bson.D{{"$set", bson.D{
+				{"promulgated-url", entity.URL},
+				{"promulgated-revision", entity.URL.Revision},
+			}}},
+		)
+		// Ignore not found errors as these will occur if the update is happening in parallel.
+		if err != nil && err != mgo.ErrNotFound {
```

Comment:
It seems to me that the "not found" case is quite important here, as this is where we can end up deleting entities if there happen to be any promulgated entities that don't have equivalent user charms. Perhaps if there's a not-found error, we could make sure that we can find the user charm, and abort if we can't, because in that case there's a real problem.